### PR TITLE
chore: replace from with_capacity with try_reserve

### PR DESCRIPTION
**Pull request by appflowy**
Replace the use of `with_capacity` with `try_reserve` in collections specifically when decoding the `Update` to better handle potential memory allocation errors. For now, this change should be applied only in the context of `Update` decoding, without altering other instances of `with_capacity` usage. 

If needed, please let me know, and I can handle the remaining tasks

### CODE REVIEW

**Comment by Horusiath** [comment]
This is a little bit tricky. A fact that you've allocated enough space for blocks alone doesn't mitigate all the issues. You can have enough memory to allocate blocks, but still not enough to allocate block-related data.

**Comment by appflowy** [comment]
Yes, refactoring is necessary for managing block-related data more effectively. In the current setup, decoding invalid data leads to a memory allocation error, which results in the immediate termination of the program. One approach I'm considering to prevent this issue is implementing 'Input Validation.' This would involve checking the data before decoding to ensure it adheres to the expected format and constraints, thereby avoiding the decoding of manifestly invalid data. Do you think this would be an effective solution, or do you have any other recommendations?

File: yffi/src/lib.rs [hunk 1]
```diff
@@ -1175,12 +1175,16 @@ pub const ERR_CODE_INVALID_JSON: u8 = 5;
 /// Error code: other error type than the one specified.
 pub const ERR_CODE_OTHER: u8 = 6;
 
+/// Error code: allocated buffer is too small to hold a decoded value.
+pub const ERR_NOT_ENOUGH_BYTES: u8 = 7;
```
Comment (Horusiath) [inline yffi/src/lib.rs:1179]: Maybe `ERR_NOT_ENOUGH_MEMORY`?

File: yrs/src/encoding/read.rs [hunk 2]
```diff
@@ -12,6 +13,9 @@ pub enum Error {
     #[error("while reading, an unexpected value was found")]
     UnexpectedValue,
 
+    #[error("failed to allocate memory: {0}")]
+    NotEnoughBytes(#[from] TryReserveError),
```
Comment (Horusiath) [inline yrs/src/encoding/read.rs:17]: Same here, `NotEnoughMemory` seems to be better description of the problem.

**Review by Horusiath — CHANGES_REQUESTED** [review]
If this is an approach to fight with malicious or corrupted updates, then we'd need to specify what is a max allowed update size and verify it after decoding the clients/block length.

But since it's hard to determine these limits before reading the entire update, the proactive approach with `try_reserve` seems to be plausible. My only note is an error name - please change it to better indicate that it's a memory allocation problem.

**Comment by appflowy** [comment]
> If this is an approach to fight with malicious or corrupted updates, then we'd need to specify what is a max allowed update size and verify it after decoding the clients/block length.
> 
> But since it's hard to determine these limits before reading the entire update, the proactive approach with `try_reserve` seems to be plausible. My only note is an error name - please change it to better indicate that it's a memory allocation problem.

Got it. Will do later

**Comment by appflowy** [comment]
@Horusiath Is there anything else I need to do for this pull request before it can be merged?

**Comment by Horusiath** [comment]
@appflowy I see there are merge conflicts. I can take it over from here and merge it this week.

**Comment by appflowy** [comment]
> @appflowy I see there are merge conflicts. I can take it over from here and merge it this week.

Done

**Review by Horusiath — APPROVED** [review]

**Pull request merged**